lemminx-liberty/src/main/java/io/openliberty/tools/langserver/lemminx/LibertyDiagnosticParticipant.java

```diff
@@ -679,4 +687,122 @@ private static void validateVariableExists(DOMDocument domDocument, List<Diagnos
             }
         }
     }
+
+    /**
+     * Validates feature compatibility by checking if features share common platforms
+     * within each platform type (Java EE, Jakarta EE, MicroProfile).
+     *
+     * @param domDocument The DOM document (server.xml)
+     * @param diagnosticsList List to add diagnostics to
+     */
+    private void validateFeatureCompatibility(Set<String> versionedFeatures, DOMDocument domDocument, List<Diagnostic> diagnosticsList) {
+        if (versionedFeatures == null || versionedFeatures.size() <= 1) {
+            return; // Need at least two features to check compatibility
+        }
+
+        // Get Liberty version and runtime from the document
+        LibertyRuntime runtimeInfo = LibertyUtils.getLibertyRuntimeInfo(domDocument);
+        String libertyVersion = runtimeInfo == null ? null : runtimeInfo.getRuntimeVersion();
+        String libertyRuntime = runtimeInfo == null ? null : runtimeInfo.getRuntimeType();
+        int requestDelay = SettingsService.getInstance().getRequestDelay();
+
+        // Maps to track features by platform type
+        Map<String, Map<String, Object>> javaEEFeatures = new HashMap<>();
+        Map<String, Map<String, Object>> jakartaEEFeatures = new HashMap<>();
+        Map<String, Map<String, Object>> microProfileFeatures = new HashMap<>();
+
+        // Find the featureManager element to get feature nodes if available
+        DOMNode featureManagerNode = LibertyUtils.getFeatureManagerElement(domDocument);
+        Map<String, DOMNode> featureNodes = featureManagerNode != null ?
+                LibertyUtils.getAllFeatureNodes(featureManagerNode) :
+                new HashMap<>();
+
+        // Categorize features by platform type
+        for (String feature : versionedFeatures) {
+            Set<String> platforms = FeatureService.getInstance().getAllPlatformsForFeature(feature, libertyVersion, libertyRuntime, requestDelay, domDocument.getDocumentURI());
+            DOMNode featureNode = featureNodes.getOrDefault(feature, null);
```

Comment:
versionedFeatures has feature names in all lower case. The featureNodes Map I believe has the original case for the features as they appeared in the file. This means you can get null here even though the feature is present. I am not sure why you have this Map when you could have passed in the feature nodes and just iterated through them.